File: structy/BinaryTree/all_tree_paths.py
class Node:
  def __init__(self, val):
    self.val = val
    self.left = None
    self.right = None

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f

def all_tree_paths(root):
    if not root:
        return []

    if not root.right and not root.left:
        return [[ root.val ]]

    paths = []

    left_paths = all_tree_paths(root.left)
    for path in left_paths:
        paths.append([ root.val, *path ])

    right_paths = all_tree_paths(root.right)
    for path in right_paths:
        paths.append([ root.val, *path ])
    
    return paths


print(all_tree_paths(a)) # ->
# [ 
#   [ 'a', 'b', 'd' ], 
#   [ 'a', 'b', 'e' ], 
#   [ 'a', 'c', 'f' ] 
# ] 

File: structy/BinaryTree/allTreePaths.js
// Write a function, allTreePaths, that takes in the root of a binary tree. The function should return a 2-Dimensional 
// array where each subarray represents a root-to-leaf path in the tree.

// The order within an individual path must start at the root and end at the leaf, but the relative order among paths in the outer 
// array does not matter.

// You may assume that the input tree is non-empty.

class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

function allTreePaths(root) {
    if (!root) return []
    if (!root.right && !root.left) return [[ root.val ]]

    const paths = []

    const rightPaths = allTreePaths(root.right)
    for (const path of rightPaths) {
        paths.push([ root.val, ...path ])
    }

    const leftPaths = allTreePaths(root.left)
    for (const path of leftPaths) {
        paths.push([ root.val, ...path ])
    }

    return paths
}

console.log(allTreePaths(a)); // ->
// [ 
//   [ 'a', 'b', 'd' ], 
//   [ 'a', 'b', 'e' ], 
//   [ 'a', 'c', 'f' ] 
// ] 
